### PR TITLE
Refactor Code Part 1

### LINKED ISSUE #15: Refactor code
Similar to #14, study the code and refactor accordingly

### PR DESCRIPTION
**Pull request by qur786**

### CODE REVIEW

**Comment by qur786** [comment]
@maneetgoyal sir Similar to #14, there was only code snippet that I have refactored. should I look through small refactoring as well in this PR? or the changes made here is enough.

File: src/calendar-heatmap.component.js [hunk 1]
```diff
@@ -158,34 +158,31 @@ export class CalendarHeatmap extends Component {
...
+                value: item.value,
+              };
+            } else {
+              summary[item.name].value += item.value;
+            }
+          });
+        }
+        return summary;
+      }, {});
+      return Object.values(summaryDictionary).sort((a, b) => {
```
Comment (maneetgoyal) [inline src/calendar-heatmap.component.js]: try to simplofy further

**Comment by maneetgoyal** [comment]
> @maneetgoyal sir Similar to #14, there was only code snippet that I have refactored. should I look through small refactoring as well in this PR? or the changes made here is enough.

more refactroing can be done by new PRs. the code look complicated and you should try to simplify it

File: src/calendar-heatmap.component.js [hunk 2]
```diff
@@ -158,34 +158,31 @@ export class CalendarHeatmap extends Component {
       'year'
     );
 
+    const getSummary = (date) => {
+      // Create 'summary' dictionary: Record<string, {name: string; value: number}>
+      let summaryDictionary = this.props.data.reduce((summary, d) => {
+        if (moment(d.date).year() === date.year()) {
+          d.summary.map((item) => {
```
Comment (maneetgoyal) [inline src/calendar-heatmap.component.js]: STudy about map vs forEach and use whats more apt

File: src/calendar-heatmap.component.js [hunk 3]
```diff
@@ -158,34 +158,31 @@ export class CalendarHeatmap extends Component {
       'year'
     );
 
+    const getSummary = (date) => {
+      // Create 'summary' dictionary: Record<string, {name: string; value: number}>
+      let summaryDictionary = this.props.data.reduce((summary, d) => {
+        if (moment(d.date).year() === date.year()) {
```
Comment (maneetgoyal) [inline src/calendar-heatmap.component.js]: Considering this line, see if `.filter().reduce()` should be done instead in line 163

File: src/calendar-heatmap.component.js [hunk 4]
```diff
@@ -158,34 +158,33 @@ export class CalendarHeatmap extends Component {
       'year'
     );
 
+    const getSummary = (date) => {
```
Comment (maneetgoyal) [inline src/calendar-heatmap.component.js]: See if you can declare this function in utils

Comment (qur786) [reply src/calendar-heatmap.component.js]: @maneetgoyal sir,
1. This function is only meant to be used for `globalOverview` which filters data according to years and output an integrated summary, other that it don't have any uses. We are not using it in any other places. 
2. We are using class specific variables `this.props.data` inside it which completely makes it class dependent. 
So in my conclusion we shouldn't do it.

Comment (maneetgoyal) [reply src/calendar-heatmap.component.js]: Whta is preventing you from passing this.props.data as input argument

Comment (qur786) [reply src/calendar-heatmap.component.js]: What about the usability sir , it is not being used anywhere.

Comment (maneetgoyal) [reply src/calendar-heatmap.component.js]: prefer maintainability more, move it

File: src/utils.js [hunk 5]
```diff
@@ -31,3 +31,30 @@ export function createColorGenerator(min_value, max_value, color) {
   }
   return colorGenerator;
 }
+
+export function getYearSummary(data, date) {
```
Comment (maneetgoyal) [inline src/utils.js:72]: Make a similar change for #14 as well, same PR

File: src/utils.js [hunk 6]
```diff
@@ -31,3 +31,56 @@ export function createColorGenerator(min_value, max_value, color) {
...
+    .flatMap((e) => e.summary)
+    .reduce((summary, item) => {
+      if (summary[item.name] === undefined) {
+        summary[item.name] = {
+          name: item.name,
+          value: item.value,
+        };
+      } else {
+        summary[item.name].value += item.value;
+      }
```
Comment (maneetgoyal) [inline src/utils.js]: create a. new function that creates this ductionary, reuse the ficntion

**Review by maneetgoyal — CHANGES_REQUESTED** [review]
chabge needed

File: src/utils.js [hunk 7]
```diff
@@ -31,3 +31,53 @@ export function createColorGenerator(min_value, max_value, color) {
...
+    .flatMap((e) => e.summary);
+  const summaryDictionary = createSummaryDictionary(summaries);
+  return sortSummaryDictionary(summaryDictionary);
+}
+
+// Calculate daily summary if that was not provided
+export function parseData(data) {
+  if (Array.isArray(data)) {
+    if (data[0].summary === null || data[0].summary === undefined) {
+      data.forEach((d) => {
```
Comment (maneetgoyal) [inline src/utils.js:61]: refactor to use map, and then update this.props.data using the fucntion output

File: src/utils.js [hunk 8]
```diff
@@ -31,3 +31,53 @@ export function createColorGenerator(min_value, max_value, color) {
   }
   return colorGenerator;
 }
+
+// Create 'summary' dictionary: Record<string, {name: string; value: number}>
+function createSummaryDictionary(summariesOrDetails) {
+  return summariesOrDetails.reduce((summary, item) => {
```
Comment (maneetgoyal) [inline src/utils.js]:
```suggestion
  return summariesOrDetails.reduce((acc, curr) => {
```

File: src/utils.js [hunk 9]
```diff
@@ -31,3 +31,53 @@ export function createColorGenerator(min_value, max_value, color) {
...
+    .filter(
+      (e) => new Date(e.date).getFullYear() === new Date(date).getFullYear()
+    )
+    .flatMap((e) => e.summary);
+  const summaryDictionary = createSummaryDictionary(summaries);
+  return sortSummaryDictionary(summaryDictionary);
+}
+
+// Calculate daily summary if that was not provided
+export function parseData(data) {
```
Comment (maneetgoyal) [inline src/utils.js]:
```suggestion
export function addSummary(data) {
```

File: src/utils.js (hunk 7 above)
Comment (maneetgoyal) [reply src/utils.js:61]: i mean `.map(...)`

Comment (qur786) [reply src/utils.js:61]: how the map function output will be used to update `data`, for that we would have to use another loop to iterate over data and update `summary` one my one on all `data elements`.

Comment (maneetgoyal) [reply src/utils.js:61]: simpler solution may be possible, think harder

File: src/utils.js [hunk 10]
```diff
@@ -31,3 +31,60 @@ export function createColorGenerator(min_value, max_value, color) {
...
+        // Create project dictionary: Record<string, {name: string; value: number}>
+        const summaryDictionary = createSummaryDictionary(d.details);
+        // Update "summary" property of the array element
+        const summary = sortSummaryDictionary(summaryDictionary);
+        return summary;
+      });
+    }
+  }
+  return summaries;
+}
```
Comment (qur786) [inline src/utils.js]: is this acceptable sir ?

Comment (maneetgoyal) [reply src/utils.js]: i don't think so

how will attach the summary object to the corresponding datum object? this approach might make it difficult

**Pull request merged**